src/routes/boot.ts

```diff
@@ -328,6 +341,49 @@ const getUser = (con: DataSource, userId: string): Promise<User> =>
     ],
   });
 
+const getMarketingCta = async (
+  con: DataSource,
+  log: FastifyBaseLogger,
+  userId: string,
+) => {
+  if (!userId) {
+    log.info('no userId provided when fetching marketing cta');
+    return null;
+  }
+
+  let marketingCta: MarketingCta | null = null;
+  const redisKey = generateStorageKey(
+    StorageTopic.Boot,
+    StorageKey.MarketingCta,
+    userId,
+  );
+  const rawRedisValue = await getRedisObject(redisKey);
+
+  if (rawRedisValue === RedisMagicValues.SLEEPING) {
+    return null;
+  } else {
+    marketingCta = JSON.parse(rawRedisValue);
+  }
+
+  if (!marketingCta) {
+    const userMarketingCta = await con.getRepository(UserMarketingCta).findOne({
+      where: { userId, readAt: IsNull() },
+      order: { createdAt: 'ASC' },
+      relations: ['marketingCta'],
+    });
+
+    if (userMarketingCta) {
+      marketingCta = userMarketingCta.marketingCta;
+      await setRedisObject(redisKey, JSON.stringify(marketingCta));
+    } else {
+      marketingCta = null;
+      await setRedisObject(redisKey, RedisMagicValues.SLEEPING);
```

Comment:
Why do we need this `SLEEPING` value? Is it not the same as no value?

Reply:
AFAIK to avoid having to re-query when empty.
But if we somehow can detect just if redis obj exist we could bypass this?

(Open to discussion)

Reply:
Empty redis key returns `''`, and `null` if it doesn't exist.

My idea was for it to indicate that there is no CTA available, and no need to actually query the database on future requests (and when there is a CTA available, we'll just override the redis key directly).

There is [`exists`](https://redis.github.io/ioredis/classes/Redis.html#exists) that can be used instead perhaps. It returns `1` when the key exists and `0` if not.

Reply:
You would have to remove SLEEPING key when user gets a new CTA to avoid it not showing due to key being there?

Reply:
Sounds good to me, better to read once then to do exist and then read the value (2 redis queries).

@idoshamun what do you think?

Reply:
There's now a TTL of 7 days on the redis key, so `SLEEPING` would expire by itself and then re-fetched after expiry.

Another way that `SLEEPING` would "expire", would be in the integration with customer.io when we assign a CTA for the user, we would just (over)write the redis key when we write to `user_marketing_cta`